compare/image.c

```diff
@@ -55,6 +64,7 @@ bool read_png(const char *path, Image16 *img)
         return false;
     }
 
+    png_set_read_fn(png, fp, &read_callback);
```

Comment:
If I understand correctly, this is a workaround for the MinGW-originating libpng using a different libc (and thus a different `FILE`) from libass on Windows?

Firstly, this deserves a comment, because this code looks like it’s completely redundant. Secondly, try changing the `msystem` in the GitHub CI matrix to `UCRT64`: assuming a new-enough MSVC, I think this should get you the right libc, making this code unnecessary. (Thirdly, of course, this is just another argument for not sourcing dependencies for an MSVC build from MSYS2.)

Reply:
This has little to do with mingw vs MSVC deps and everything to do with dynamic deps vs static deps. See [this Microsoft docs article](https://learn.microsoft.com/en-us/cpp/c-runtime-library/potential-errors-passing-crt-objects-across-dll-boundaries?view=msvc-170): on Windows, it should be presumed unsafe to pass libc structures across DLL boundaries, or to e.g. `malloc()` an object in one DLL and `free()` it in another. Dynamically-linked deps aren't uncommon by any means, and IMO are a case worth having tests for in our CI.

There are two related potential issues here:
- Different libraries being linked against different *versions* of the libc code (mingw vs MSVC, but also eg MSVC 2013 vs 2015, or UCRT vs MSVCRT)
- Different libraries calling into different *copies* of the same libc code (e.g. either [or both] are statically linked against UCRT)

In either of these cases (which are very normal, esp since statically linking the libc is common on windows, and indeed is enabled by default in some situations), the DLLs have separate heaps, and it is not safe to pass libc/CRT objects (including `FILE*`) across the DLL boundary. This is why we have #775, and why it's so common for libraries to have similar routines (ffmpeg, openssl, etc). This is a common error, and seems extremely worth having tests for!

Reply:
> Different libraries being linked against different versions of the libc code (mingw vs MSVC, but also eg MSVC 2013 vs 2015, or UCRT vs MSVCRT)

That's historically true, and for that reason many binary releases are/were shipping multiple .dll versions built with each of the CRT. But Microsoft resolved those issue with UCRT. They no longer break compatibility on each version.

It's even noted in the article linked

> Because the DLL version of the Universal CRT library used by Visual Studio 2015 and later is now a centrally deployed Windows component (ucrtbase.dll), it's the same for apps built with Visual Studio 2015 and later versions. 

I strongly believe that in present year, there is no need to care about MSVC 2013 (EOL Apr 9, 2019) or MSVC 2015 (EOL Oct 13, 2020) or about MSVCRT at all. Everyone should be using UCRT which is available since Windows Vista SP2 (with an update). And if anyone needs to deploy MSVCRT they should likely built their dependencies to avoid any issues.

MinGW switched by default to UCRT, which in turn is compatible with MSVC builds.

(the only reason MSVCRT is still popular is because prebuilt toolchains in linux distros are outdated)

DLLs on windows are self-contained, unlike in POSIX world symbols are not loaded into global namespaces and by design thy can use different libc or have it duplicated.

> This is a common error, and seems extremely worth having tests for!

Sure. For me this is Frankenstein build, it doesn't reflect real scenario. And if we want to test corner cases, we should also have nominal cases tested, otherwise what is the point?

Reply:
I didn’t expect that this affects static builds even when using the same libc, but it makes sense.

Anyway, I’d appreciate a short code comment here explaining this is done to ensure the same stdio is called to open the file and to read it regardless of what stdio libpng is linked to. (Technically, this also affects non-Windows; it’s just more rare.)

> MinGW switched by default to UCRT, which in turn is compatible with MSVC builds.

Not sure what this means. MSYS2 has multiple environments with different C libraries (all of them with MinGW-w64 compilers), and you choose one of them explicitly. This is what I suggested changing in the CI configuration.

Reply:
> Not sure what this means. MSYS2 has multiple environments with different C libraries (all of them with MinGW-w64 compilers), and you choose one of them explicitly. This is what I suggested changing in the CI configuration.

MSYS2 is not MinGW. MSYS2 is package manager. They provide various environments. See, https://www.msys2.org/docs/environments/ especially MSVCRT vs UCRT section.

Anyway, what I meant is that [MinGW](https://www.mingw-w64.org/) starting from v12.0.0 is using UCRT by default. Of course you can build MSVCRT version (like MSYS2 provides), but we are at the point where we should be migrating everything to UCRT.

See also https://sourceforge.net/p/mingw-w64/mingw-w64/ci/master/tree/mingw-w64-doc/howto-build/ucrt-vs-msvcrt.txt

> This is what I suggested changing in the CI configuration.

This is good suggestion, also worth testing is CLANG64 environment.

Reply:
Added comment clarifying intent.

Reply:
reminder: `test` writes pngs and so does `compare`. If we consider this a supported usecase this needs fixing too and if not this is change can be removed once we stop mixing in CI

Reply:
Yes, that should be added; I considered it out-of-scope for this PR (esp since we don't have any tests that use those, so it'd be hard to validate).

Reply:
> esp since we don't have any tests that use those, so it'd be hard to validate

The muon CI job actually builds and executes the `test` utility atm though it ofc doesn't use MSVC. With the PR here it now makes more sense to run the usual tests instead